interpreter/cling/lib/MetaProcessor/MetaSema.cpp

```diff
@@ -466,7 +466,7 @@ namespace cling {
       clang::ASTContext& Ctx = m_Interpreter.getCI()->getASTContext();
       if (result) {
         *result = Value(Ctx.IntTy, m_Interpreter);
-        result->getAs<long long>() = exitStatus;
+        result->setLongLong(exitStatus); // FIXME: This should assert.
```

Comment:
Yes, that's a mistake here (and in the end the result is use exactly once in the code to compare the value to an enum, so why not change the type now :) )  but why would you prevent `setLongLong` from taking a smaller-width integer as argument?  I.e. why would you assert there anymore than the compiler is warning here:
```
long long value = 1u;
```

(and of course in the reverse situation, it can warn:
```
root [2] int i = 1ll << 33
ROOT_prompt_2:1:13: warning: implicit conversion from 'long long' to 'int' changes value from 8589934592 to 0 [-Wconstant-conversion]
int i = 1ll << 33
    ~   ~~~~^~~~~
```
)

Reply:
> Yes, that's a mistake here (and in the end the result is use exactly once in the code to compare the value to an enum, so why not change the type now :) ) 

Because I will do that in a separate commit where I enable the assert. This way it would have clearer history as that's not the only place we do that type of "mistake". Here the "right" use would be using `cling::Value::Create(..., exitStatus)` and we will get the correct underlying type.

> but why would you prevent `setLongLong` from taking a smaller-width integer as argument? I.e. why would you assert there anymore than the compiler is warning here:

Sure, we could probably get away in assigning a smaller-width compatible types. Allowing this while catching the other cases will slow down the implementation of `setX` and it was meant to be as fast as a memory access. Then the question becomes about maintaining consistency between the type and its value. When you have a `cling::Value` with underlying type of `long long` would setting an `int` mean also adjust the type? The right idiom to use here is `cling::Value::Create` which will make the thing.

It is still an open question probably for further work do we want `cling::Value` to be multiply updated by user code. So far most of the use patterns I've seen are create a `cling::Value`, set its value and pass it as an input parameter; or create a `cling::Value` and send it to the JIT to put the output result there. I see a little sense for these things to be mutable especially by user code...

Reply:
I missed that the comment was about the discrepancies between line 468:
```
*result = Value(Ctx.IntTy, m_Interpreter);
```
and line 469
```
result->setLongLong(exitStatus); 
```
i.e. I missed the `Ctx.IntTy` part and instead was focusing on the difference between `setLongLong` and the `typeof(exitStatus)`.  The later I still argue is okay (otherwise you would have to assert on `setLongLong( 1 );` !!)
But the earlier I agree with you is wrong: i.e. it breaks the contract that we were talking about earlier: "I know the type hence I can use the fast interface".

> Sure, we could probably get away in assigning a smaller-width compatible types. 

Now that I understand that we are talking about 'registered type of the Value' vs 'function called to set it', I agree that this should not be supported (it may or may not have the intended behavior depending on endianess, etc.).

> Because I will do that in a separate commit where I enable the assert. 

Fair enough.

Reply:
> i.e. I missed the Ctx.IntTy part and instead was focusing on the difference between setLongLong and the typeof(exitStatus). The later I still argue is okay (otherwise you would have to assert on setLongLong( 1 ); !!)

`setLongLong` will take a `long long` and the compiler would convert `1` to `long long`. Unless the `cling::Value`'s type is not `LongLongTy` we will not assert.